core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java

```diff
@@ -59,18 +65,48 @@ protected final EndpointGroup group() {
     /**
      * Initialize this {@link EndpointSelector} to listen to the new endpoints emitted by the
      * {@link EndpointGroup}. The new endpoints will be passed to {@link #updateNewEndpoints(List)}.
+     *
+     * <p>This method is called automatically when the first selection is made. However, if you want to
+     * start listening to the {@link EndpointGroup} earlier, you can call this method manually.
      */
-    @UnstableApi
     protected final void initialize() {
-        endpointGroup.addListener(this::refreshEndpoints, true);
+        tryInitialize();
+    }
+
+    private void tryInitialize() {
+        if (initialized == 1) {
+            return;
+        }
+
+        if (initializedUpdater.compareAndSet(this, 0, 1)) {
+            endpointGroup.addListener(this::refreshEndpoints, true);
+        }
+    }
+
+    @VisibleForTesting
+    boolean isInitialized() {
+        return initialized == 1;
     }
 
     @Override
     public CompletableFuture<Endpoint> select(ClientRequestContext ctx, ScheduledExecutorService executor,
                                               long timeoutMillis) {
+        tryInitialize();
```

Comment:
Question) For my understanding, it seems reasonable to assume that pretty much all cases of extending `AbstractEndpointSelector` will want to listen to the `endpointGroup`.
Why isn't `tryInitialize()` called from the constructor? Was there a race condition?

Reply:
Good question. `tryInitialize()` may invoke `updateNewEndpoints` synchronously if the `endpointGroup` has been updated with the initial endpoints. Since `updateNewEndpoints` can access member fields of subclasses, calling `tryInitialize()` in the constructor of `AbstractEndpointSelector` may result in `updateNewEndpoints` being invoked before those  fields are initialized.
https://github.com/line/armeria/blob/afa61ca10e25c861c34de86b821d7d2b38bba7cd/core/src/main/java/com/linecorp/armeria/client/endpoint/DefaultEndpointSelector.java#L37-L40
https://github.com/line/armeria/blob/afa61ca10e25c861c34de86b821d7d2b38bba7cd/core/src/main/java/com/linecorp/armeria/client/endpoint/DefaultEndpointSelector.java#L58-L64